crates/handlers/src/oauth2/registration.rs

```diff
@@ -111,55 +123,21 @@ pub(crate) async fn post(
 ) -> Result<impl IntoResponse, RouteError> {
     info!(?body, "Client registration");
 
-    // Let's validate a bunch of things on the client body first
-    for uri in &body.redirect_uris {
-        if uri.fragment().is_some() {
-            return Err(RouteError::InvalidRedirectUri);
-        }
-    }
-
-    // Check that the client did not send both a jwks and a jwks_uri
-    if body.jwks_uri.is_some() && body.jwks.is_some() {
-        return Err(RouteError::InvalidClientMetadata);
-    }
-
-    // Check that the grant_types and the response_types are coherent
-    let has_implicit = body.grant_types.contains(&GrantType::Implicit);
-    let has_authorization_code = body.grant_types.contains(&GrantType::AuthorizationCode);
-    let has_both = has_implicit && has_authorization_code;
-
-    for response_type in &body.response_types {
-        let is_ok = match response_type {
-            OAuthAuthorizationEndpointResponseType::Code => has_authorization_code,
-            OAuthAuthorizationEndpointResponseType::CodeIdToken
-            | OAuthAuthorizationEndpointResponseType::CodeIdTokenToken
-            | OAuthAuthorizationEndpointResponseType::CodeToken => has_both,
-            OAuthAuthorizationEndpointResponseType::IdToken
-            | OAuthAuthorizationEndpointResponseType::IdTokenToken
-            | OAuthAuthorizationEndpointResponseType::Token => has_implicit,
-            OAuthAuthorizationEndpointResponseType::None => true,
-        };
-
-        if !is_ok {
-            return Err(RouteError::InvalidClientMetadata);
-        }
-    }
-
-    // If the private_key_jwt auth method is used, check that we actually have a
-    // JWKS for that client
-    if body.token_endpoint_auth_method == Some(OAuthClientAuthenticationMethod::PrivateKeyJwt)
-        && body.jwks_uri.is_none()
-        && body.jwks.is_none()
-    {
-        return Err(RouteError::InvalidClientMetadata);
-    }
+    // Validate the body
+    let metadata = body.validate()?;
 
     let mut policy = policy_factory.instantiate().await?;
-    let res = policy.evaluate_client_registration(&body).await?;
+    let res = policy.evaluate_client_registration(&metadata).await?;
     if !res.valid() {
         return Err(RouteError::PolicyDenied(res.violations));
     }
 
+    // Contacts was checked by the policy
+    let contacts = match metadata.contacts.as_deref() {
+        Some(contacts) => contacts,
+        None => unreachable!(),
+    };
```

Comment:
Why isn't this one a getter/method in `VerifiedClientMetadata` like the other ones?

Reply:
Because it's not mandatory in OpenID Connect, so it's only enforced by the OPA policy.

Reply:
Makes sense! :)
Although, the HS admin might run the server with a policy which might not require the contacts field, so I'd rather default to an empty slice here instead of crashing

Reply:
That's right I was wondering if the policy could be changed and I forgot to ask you.